src/os_tips.eliom

```diff
@@ -189,12 +189,101 @@ let%shared block ?(a = []) ?(recipient = `All)
-let%client waiter = ref (onload_waiter ())
+(* This list of threads is used to display only one tip at a time,
+   in the order specified by the given priorities *)
+let%client prioritized_waiters : (int option * unit Lwt.t * bool) list ref =
+  ref []
+
+(* [Lwt.cancel] does nothing if the task is already resolved,
+   so we can safely cancel them all *)
+let%client cancel_waiters () =
+  List.iter (fun (_,w,_) -> Lwt.cancel w) !prioritized_waiters
+
+(* This boolean is used to track whether
+   the list of priorities has been sorted
+ *)
+let%client sorted = ref false
+
+(* A priority of [None] is considered infinite,
+   and thus greater than everything else *)
+let%client compare_priority_opt p1 p2 =
+  match p1,p2 with
+  | None, None -> 0
+  | None, _ -> 1
+  | Some _, None -> -1
+  | Some p1, Some p2 -> compare p1 p2
+
+(* Find the appropriate promise to wait for
+   corresponding to the given priority.
+
+   This function assumes the input list is sorted.
+
+   It turns to 'true' any priority item matched with its previous waiter.
+   That much is useful to keep track of which items to ignore.
+
+   It returns [None] if no appropriate promise is found.
+   Otherwise it returns [Some (promise,l)]
+   where [l] is meant to replace [prioritized_waiters]
+   and [promise] is the promise the bubble calling the function should wait for.
+ *)
+let%client rec find_previous priority = function
+  | [] -> (* Not found in an empty list *) None
+  | (p,_,_)::_ when compare_priority_opt p priority > 0 ->
+    (* First priority is too low: Not Found *)
+    None
+  | (p,w,false)::l when p = priority ->
+    (* Very first priority found: Result waiter resolves immediately. *)
+    Some (Lwt.return_unit, (p,w,true)::l)
+  | (prevp, prevw, prevb)::(p,w,false)::tl
+       when p = priority ->
+    (* First of a series of priorities is available:
+       Result waiter is the previous in the queue *)
+    Some (prevw, (prevp,prevw,prevb)::(p,w,true)::tl)
+  | (prevp, prevw, prevb)::(p,w,true)::tl
+       when p = priority ->
+    (* First of a series of priorities is unavailable:
+       Keep looking and rebuild on top of the list *)
+    (match find_previous priority ((p,w,true)::tl) with
+     | None -> None
+     | Some (r,l) -> Some (r, (prevp,prevw,prevb)::l))
+  | (p,w,b)::l
+       when compare_priority_opt p priority < 0 ->
+    (* Following priority is not matched by previous cases:
+       Keep looking and rebuild on top of the list *)
+    (match find_previous priority l with
+     | None -> None
+     | Some (r,l) -> Some (r, (p,w,b)::l))
+  | _ ->
+    (* Catch-all because everything else uses guards,
+       but should be unreachable *)
+    assert false
+
+let%client wait_for_bubble ?priority () =
+  (* We wait for the elements to load,
+     to be sure we have all waiters prioritized *)
+  let%lwt () = onload_waiter () in
+  if not !sorted then
+    (prioritized_waiters := List.rev !prioritized_waiters; sorted := true);
```

Comment:
Why not sorting in reverse order directly?

Reply:
It would reverse the call-order too. On priority equality, first calls are inserted first, thus later in the list. We have no other reliable way to keep this order than reversing the list.

Reply:
ok thanks

Reply:
That said, I'm currently sorting the list on every addition, which is suboptimal. I'll move the sort to this place so we only do it once (per batch)